### PR TITLE
Spark: Add Spark extension for table encryption key

### PR DESCRIPTION
**Pull request by huaxingao**
This is the first PR for supporting table encryption key syntax proposed in this https://docs.google.com/document/d/1kkcjr9KrlB9QagRX3ToulG_Rf-65NMSlVANheDNzJq4/edit#heading=h.5e5qu1qyfmhm

The new syntax we will support:
```
ALTER TABLE table ENCRYPTED WITH KEK|MEK BY TABLE KEY key-id
```
This PR added the new syntax and the new logical/physical operators. Currently the run method in the physical operator will throw `UnsupportedOperationException`. Next PR will implement the run method to generate the table encryption key  and change the table into an encrypted table.

### CODE REVIEW

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical
+
+import org.apache.spark.sql.catalyst.expressions.Attribute
+
+case class CreateTableEncKey(
```
Comment (kbendick) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala:24]: Nit / non-blocking: In the Spark repo, are these normally termed with `Command` at the end (like `CreateTableCommand`, etc)?

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    table: Seq[String],
```
Comment (kbendick) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala:25]: Is there a reason to prefer `Seq[String]` here? Would it possibly make more sense to use an identifier type here, such as `org.apache.spark.sql.connector.catalog.Identifier`?

If other ones use `Seq[String]` or if there's a reason that we can't use a more definitive type, feel free to close this comment =)

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  @After
+  public void removeTable() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testEncKey() {
+
+    sql("CREATE TABLE %s (id bigint NOT NULL, " +
+        "location struct<lon:bigint NOT NULL,lat:bigint NOT NULL> NOT NULL) USING iceberg", tableName);
```
Comment (kbendick) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java:43]: Nit: Is it possible to use a different field name other than `location` in the schema? Maybe `coordinates`?

At first I thought that this was a reference to the `LOCATION` keyword for create table DDL etc and was somewhat confused.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala (hunk 2 above)
Comment (kbendick) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala:25]: EDIT: I took a look, and it seems that `Seq[String]` is the standard way of doing it in some of the others as well (plus you use spark catalog `Identifier` class in the implementation.

Please feel free to resolve these comments =)

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala [hunk 4]
```diff
@@ -115,6 +116,18 @@ class IcebergSqlExtensionsAstBuilder(delegate: ParserInterface) extends IcebergS
       ctx.fieldList.fields.asScala.map(_.getText))
   }
 
+  /**
+   * Create a ENCRYPTED WITH (KEK | MEK) BY TABLE KEY command.
```
Comment (kbendick) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:120]: Nit: You might want to reference the class that's created in the Scala doc, like the one below here.

Something like `Create a [[CreateTableEncKeyCommand]] for creating (applying?) a KEK or MEK encryption key for the table`.

The actual language is up to you as you, but it would be nice if the Scala Docs referenced the command class using `[[...]]`. 🙂

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala (hunk 1 above)
Comment (huaxingao) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala:24]: Right, these normally termed with `Command` at the end in Spark repo, but seems all the logical commands in `spark3-extensions` under `org.apache.spark.sql.catalyst.plans.logical` don't have `Command` in the end, e.g. `SetIdentifierFields`, so I guess I will just be consistent with others.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala (hunk 2 above)
Comment (huaxingao) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/CreateTableEncKey.scala:25]: I will leave this as is :)

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java (hunk 3 above)
Comment (huaxingao) [reply spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java:43]: Modified. Thanks!

File: spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala (hunk 4 above)
Comment (huaxingao) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/parser/extensions/IcebergSqlExtensionsAstBuilder.scala:120]: Fixed. Thanks!
The majority of the visitXXX methods in this class don't use the `[[...]]`, though. Should I fix those too?

**Comment by huaxingao** [comment]
cc @ggershinsky @jackye1995
The proposed syntax is  `ALTER TABLE table ENCRYPTED WITH KEK|MEK BY DATAFILE KEY key-id` in https://docs.google.com/document/d/1kkcjr9KrlB9QagRX3ToulG_Rf-65NMSlVANheDNzJq4/edit#heading=h.5e5qu1qyfmhm, but this was changed to `ALTER TABLE table ENCRYPTED WITH KEK|MEK BY TABLE KEY key-id` per Gidon's suggestion.

File: spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 5]
```diff
@@ -73,6 +73,7 @@ statement
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
     | ALTER TABLE multipartIdentifier SET IDENTIFIER_KW FIELDS fieldList                    #setIdentifierFields
     | ALTER TABLE multipartIdentifier DROP IDENTIFIER_KW FIELDS fieldList                   #dropIdentifierFields
+    | ALTER TABLE multipartIdentifier ENCRYPTED WITH (KEK | MEK) BY TABLE KEY keyId         #createTableEncKey
```
Comment (flyrain) [inline spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: 1. Are we going to add `BY COLUMN` later to support column level encryption? 
2. Maybe we don't need "BY TABLE", the key applies to all columns by default.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateTableEncKeyExec.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  override lazy val output: Seq[Attribute] = Nil
+
+  override protected def run(): Seq[InternalRow] = {
+    catalog.loadTable(ident) match {
+      case iceberg: SparkTable =>
+        // Todo: This will be replaced by Table encryption key implementation in next PR
+        throw new UnsupportedOperationException(s"create Table encryption key not supported yet: " +
+          s"${catalog.name}.${ident.quoted} isKEK: $isKEK $keyId")
+      case table =>
+        throw new UnsupportedOperationException(s"Cannot create Table encryption key: $table")
```
Comment (flyrain) [inline spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateTableEncKeyExec.scala:44]: like this "Cannot create Table encryption key for a non-Iceberg table"?

File: spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 5 above)
Comment (flyrain) [reply spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: Is there a plan to add this to command `CREATE TABLE`?

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testEncKey() {
+
+    sql("CREATE TABLE %s (id bigint NOT NULL, " +
+        "coordinates struct<lon:bigint NOT NULL,lat:bigint NOT NULL> NOT NULL) USING iceberg", tableName);
+    Table table = validationCatalog.loadTable(tableIdent);
+    Assert.assertTrue("Table should start without identifier", table.schema().identifierFieldIds().isEmpty());
```
Comment (flyrain) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java:45]: these two lines are unrelated. Do we need them?

File: spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 5 above)
Comment (huaxingao) [reply spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: Yes, I think we will need to support `CREATE TABLE` too, and also `BY COLUMN`, but `BY COLUMN` is for phase 2B.
I agree that `BY TABLE` seems not required. Removed for now.

File: spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateTableEncKeyExec.scala (hunk 6 above)
Comment (huaxingao) [reply spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateTableEncKeyExec.scala]: Changed.

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java (hunk 7 above)
Comment (huaxingao) [reply spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTableEncryption.java]: Removed

**Comment by jackye1995** [comment]
I'd like to understand more about the reason behind adding this extension so early. My understanding is that most features needed to support this is not implemented in Iceberg yet, and Spark SQL extension is typically the last step after everything is ready.

I was pulled to first focus on the Trino and PrestoDB Iceberg connector work in the past few months so I did not follow up too much on this area, do we already have all the building pieces to make this work? If not, I would propose we focus on those aspects first before adding the SQL extension, as the syntax might change if the base implementation changes.

**Comment by ggershinsky** [comment]
Back from the vacation. This PR is a part of the "encryption package", comprised of a number of disjoint pull requests. I've sent a combined PR #3053  that creates an end-to-end implementation - sans the DDL part (using table properties for encryption activation). @huaxingao , will appreciate if you'd have a look at it for a compatibility with your PR down the road. I agree we should converge first on the implementation design and details, adding the DDL interface as the next step. Still, I think it's good to have all pieces in place before we start iterating on them.

File: spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 5 above)
Comment (ggershinsky) [reply spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4]: +1 on both accounts - adding this to `CREATE TABLE`, and replacing `BY TABLE KEY` with `BY KEY`

File: spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 8]
```diff
@@ -73,6 +73,7 @@ statement
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
     | ALTER TABLE multipartIdentifier SET IDENTIFIER_KW FIELDS fieldList                    #setIdentifierFields
     | ALTER TABLE multipartIdentifier DROP IDENTIFIER_KW FIELDS fieldList                   #dropIdentifierFields
+    | ALTER TABLE multipartIdentifier ENCRYPTED WITH (KEK | MEK) KEY keyId                  #createTableEncKey
```
[the remark this answers is not in the document]
Comment (ggershinsky) [reply spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: This is the problem with the `ALTER` syntax.. Currently, we don't have a support for converting an existing regular table to an encrypted table. We will have it in the future, but it will be an explicit and massive data-crunching procedure. For now, we focus on a simpler and more mainstream goal - creating an empty table with an encryption key, so all future data writes end up in encrypted files. It seems to me the `CREATE` syntax is a better fit here..

Regarding dropping the key(s) - I'm not aware of usecases for such functionality. It's possible to think of some, but given the ambiguity and complexity of this function, I'd suggest considering it at a later stage.

[the remark this answers is not in the document]
Comment (flyrain) [reply spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:76]: Thanks for the explanation @ggershinsky. Would suggest to drop change of `ALTER TABLE`.
